ports/esp32/machine_i2c.c

```diff
@@ -49,92 +49,159 @@
 
 #define I2C_DEFAULT_TIMEOUT_US (50000) // 50ms
 
+// ---------------- Internal data structures ----------------
 typedef struct _machine_hw_i2c_obj_t {
     mp_obj_base_t base;
-    i2c_port_t port : 8;
+    i2c_master_bus_handle_t bus_handle;
+    i2c_master_dev_handle_t dev_handle;
+    uint8_t port : 8;
     gpio_num_t scl : 8;
     gpio_num_t sda : 8;
 } machine_hw_i2c_obj_t;
 
 static machine_hw_i2c_obj_t machine_hw_i2c_obj[I2C_NUM_MAX];
 
+// ---------------- Initialization ----------------
 static void machine_hw_i2c_init(machine_hw_i2c_obj_t *self, uint32_t freq, uint32_t timeout_us, bool first_init) {
-    if (!first_init) {
-        i2c_driver_delete(self->port);
+
+    // 1. If already initialized, uninstall the old driver first
+    if (!first_init && self->bus_handle) {
+        i2c_master_bus_rm_device(self->dev_handle);
+        i2c_del_master_bus(self->bus_handle);
+        self->bus_handle = NULL;
+        self->dev_handle = NULL;
     }
-    i2c_config_t conf = {
-        .mode = I2C_MODE_MASTER,
-        .sda_io_num = self->sda,
-        .sda_pullup_en = GPIO_PULLUP_ENABLE,
+
+    // 2. Configure the bus
+    i2c_master_bus_config_t bus_cfg = {
+        .i2c_port = self->port,
         .scl_io_num = self->scl,
-        .scl_pullup_en = GPIO_PULLUP_ENABLE,
-        .master.clk_speed = freq,
+        .sda_io_num = self->sda,
+        .clk_source = I2C_CLK_SRC_DEFAULT,
+        .glitch_ignore_cnt = 7,
+        .flags.enable_internal_pullup = true,
     };
-    i2c_param_config(self->port, &conf);
-    int timeout = I2C_SCLK_FREQ / 1000000 * timeout_us;
-    i2c_set_timeout(self->port, (timeout > I2C_LL_MAX_TIMEOUT) ? I2C_LL_MAX_TIMEOUT : timeout);
-    i2c_driver_install(self->port, I2C_MODE_MASTER, 0, 0, 0);
+    ESP_ERROR_CHECK(i2c_new_master_bus(&bus_cfg, &self->bus_handle));
+
+    // 3. Add a device (placeholder address; will be changed dynamically later)
+    i2c_device_config_t dev_cfg = {
+        .dev_addr_length = I2C_ADDR_BIT_LEN_7,
+        .device_address = 0x00,               // Placeholder
+        .scl_speed_hz = freq,
+    };
+    ESP_ERROR_CHECK(i2c_master_bus_add_device(self->bus_handle, &dev_cfg, &self->dev_handle));
 }
 
 int machine_hw_i2c_transfer(mp_obj_base_t *self_in, uint16_t addr, size_t n, mp_machine_i2c_buf_t *bufs, unsigned int flags) {
     machine_hw_i2c_obj_t *self = MP_OBJ_TO_PTR(self_in);
 
-    i2c_cmd_handle_t cmd = i2c_cmd_link_create();
+    /* 0. Probe the address to see if any device responds */
+    esp_err_t err = i2c_master_probe(self->bus_handle, addr, 1000);
+    if (err != ESP_OK) {
+        return -MP_ENODEV;          /* No device at address, return immediately */
+    }
+    /* 1. Create a temporary device handle for this transaction */
+    i2c_device_config_t dev_cfg = {
```

Comment:
Is it necessary to create this temporary device handle, or should't it be better created once at instantiation?
And if, it should be set up with the configured scl freq and not a default value.

Reply:
Uüdate: It is necessary, at least with IDF V5.4.2. IDF V5.5 has an API to change the I2C address of a dev_handle. Then the handle has to be created only once.